File: android/src/main/java/expo/modules/arview/ModelConfig.kt
package expo.modules.arview

data class ModelConfig(
    val id: String = "",
    val sourceUri: String,
    val placement: String = "tap",
    val scale: Float = 1.0f,
    val rotation: FloatArray = floatArrayOf(0f, 0f, 0f),
    val gestureScale: Boolean = false,
    val gestureRotate: Boolean = false,
    val gestureScaleMin: Float = 0.1f,
    val gestureScaleMax: Float = 10.0f,
    val gestureScaleSensitivity: Float = 1.0f
) {
    companion object {
        fun fromMap(map: Map<String, Any?>): ModelConfig {
            return ModelConfig(
                id = map["id"] as? String ?: "",
                sourceUri = map["sourceUri"] as? String ?: "",
                placement = map["placement"] as? String ?: "tap",
                scale = (map["scale"] as? Number)?.toFloat() ?: 1.0f,
                rotation = (map["rotation"] as? List<*>)?.let { list ->
                    floatArrayOf(
                        (list.getOrNull(0) as? Number)?.toFloat() ?: 0f,
                        (list.getOrNull(1) as? Number)?.toFloat() ?: 0f,
                        (list.getOrNull(2) as? Number)?.toFloat() ?: 0f
                    )
                } ?: floatArrayOf(0f, 0f, 0f),
                gestureScale = map["gestureScale"] as? Boolean ?: false,
                gestureRotate = map["gestureRotate"] as? Boolean ?: false,
                gestureScaleMin = (map["gestureScaleMin"] as? Number)?.toFloat() ?: 0.1f,
                gestureScaleMax = (map["gestureScaleMax"] as? Number)?.toFloat() ?: 10.0f,
                gestureScaleSensitivity = (map["gestureScaleSensitivity"] as? Number)?.toFloat() ?: 1.0f
            )
        }
    }
}

File: ios/ModelConfig.swift
import Foundation

struct ModelConfig {
    let id: String
    let sourceUri: String
    let placement: String
    let scale: Float
    let rotation: [Float]
    let gestureScale: Bool
    let gestureRotate: Bool
    let gestureScaleMin: Float
    let gestureScaleMax: Float
    let gestureScaleSensitivity: Float

    static func from(_ dict: [String: Any]) -> ModelConfig {
        let rotation: [Float]
        if let rotList = dict["rotation"] as? [Any] {
            rotation = [
                (rotList[safe: 0] as? NSNumber)?.floatValue ?? 0,
                (rotList[safe: 1] as? NSNumber)?.floatValue ?? 0,
                (rotList[safe: 2] as? NSNumber)?.floatValue ?? 0
            ]
        } else {
            rotation = [0, 0, 0]
        }

        return ModelConfig(
            id: dict["id"] as? String ?? "",
            sourceUri: dict["sourceUri"] as? String ?? "",
            placement: dict["placement"] as? String ?? "tap",
            scale: (dict["scale"] as? NSNumber)?.floatValue ?? 1.0,
            rotation: rotation,
            gestureScale: dict["gestureScale"] as? Bool ?? false,
            gestureRotate: dict["gestureRotate"] as? Bool ?? false,
            gestureScaleMin: (dict["gestureScaleMin"] as? NSNumber)?.floatValue ?? 0.1,
            gestureScaleMax: (dict["gestureScaleMax"] as? NSNumber)?.floatValue ?? 10.0,
            gestureScaleSensitivity: (dict["gestureScaleSensitivity"] as? NSNumber)?.floatValue ?? 1.0
        )
    }
}

private extension Array {
    subscript(safe index: Int) -> Element? {
        indices.contains(index) ? self[index] : nil
    }
}
